documentation/projects/proposals/search_relevancy_sandbox/20230518-implementation_plan_staging_index_rapid_iteration.md

```diff
@@ -0,0 +1,390 @@
+schedule of `None` so that it is only run when triggered. It should also have
+`max_active_runs` set to `1` so that only one instance of the DAG can be running
+at a time.
+
+### Parameters
+
+1. `media_type`: The media type for which the index is being created. Presently
+   this would only be `image` or `audio`.
+2. `index_config`: A JSON object containing the configuration for the new index.
+   The values in this object will be merged with the existing configuration,
+   where the value specified at a leaf key in the object will override the
+   existing value (see [Merging policy](#merging-policy) below). This can also
+   be the entire index configuration, in which case the existing configuration
+   will be replaced entirely (see `override_config` parameter below).
+3. `index_suffix`: (Optional) The name suffix of the new index to create. This
+   will be a string, and will be used to name the index in Elasticsearch of the
+   form `{media_type}-{index_suffix}`. If not provided, the suffix will be a
+   timestamp of the form `YYYYMMDDHHMMSS`.
+4. `source_index`: (Optional) The existing index on Elasticsearch to use as the
+   basis for the new index. If not provided, the index aliased to `media_type`
+   will be used (e.g. `image` for the `image` media type). In production, the
+   data refresh process creates the index and aliases it to the media type. In
+   staging, the process for creating the indices which can be iterated on here
+   will be defined in a follow-up IP, #1987.
+5. `override_config`: (Optional) A boolean value which can be toggled to replace
+   the existing index configuration entirely with the new configuration. If
+   `True`, the `index_config` parameter will be used as the entire
+   configuration. If `False`, the `index_config` parameter will be merged with
+   the existing configuration. Defaults to `False`.
+6. `query`: (Optional) An
+   [Elasticsearch query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl.html)
+   to use to filter the documents to be copied to the new index. If not
+   provided, all documents will be copied. See
+   [the reindex API endpoint for the request body](https://www.elastic.co/guide/en/elasticsearch/reference/7.17/docs-reindex.html#docs-reindex-api-request-body)
+   and
+   [`create_and_populate_filtered_index` on the ingestion server](https://github.com/WordPress/openverse/blob/3fcce5ade2165955db5bbcb4f679257b3260547b/ingestion_server/ingestion_server/indexer.py#L529)
+   for an example of how this is used.
+
+### DAG
+
+Once the parameters are provided, the DAG will execute the following steps:
+
+1. (For `production` only) Check that there are no other reindexing jobs
+   currently underway. This can be done by checking whether the
+   `{media_type}_data_refresh` and `create_filtered_{media_type}_index` DAGs are
+   currently running. If either of them are, this DAG should fail immediately
+   with an appropriate error message.
```

Comment:
Should we add a note that we need to do this check in the `{media_type}_data_refresh` and `create_filtered_{media_type}_index` DAGs as well? Would it not be an issue if one of those DAGs started while this was running?

Reply:
Ah, I hadn't thought of the case where this was started before those 😖 It felt easy enough to fail this DAG in the first case, but for the other DAGs we'll need to probably have a more robust mechanism for handling this...I'll try to add some clarification/thoughts on this before this gets merged.

Reply:
The data refresh DAG already waits for the create_filtered_index DAGs to finish. We could do the same for these new ones. So long as the new DAGs are only allowed to run 1 at a time, they wouldn't be able to indefinitely block the data refresh DAG from running, especially if the new DAGs themselves fail when the data refresh is running.

Reply:
That's true! And that should be easy enough to add to the list of conditions for the data refresh. I'll do that!